File: Urnas-y-Bolas-Profe/src/Bola.java
import org.fusesource.jansi.AnsiConsole;

import static org.fusesource.jansi.Ansi.ansi;

public class Bola
{
    private static final char BOLA='\u25CF';    // Representación textual
    private static int contador = 0;            // Nº de bolas que existen
    private int numero;                         // >=1, AUTO, NO MODIFICABLE
    private Color color;                        // NO MODIFICABLE, NO NULO
    
    public Bola(Color color) {
        setNumero(++contador);
        setColor(color);
    }
    
    public static Bola generar() {
        Color[] colores = Color.values();
        int n = (int)(Math.random()*colores.length);
        return new Bola(colores[n]);
    }

    // Accesores
    
    public Color getColor() {
        return color;
    }
    public int getNumero() {
        return numero;
    }
        
    // Mutadores
    
    private void setNumero(int numero) {
        assert numero>=1: 
            String.format("El numero debe ser >=1 (numero=%d)", numero);
        this.numero = numero;
    }
    
    private void setColor(Color color) {
        assert color!=null: "El color no puede ser nulo";       
        this.color = color;        
    }
    
    public static int totalBolas() {
        return contador;
    }

    public String toString() {
        return ansi().fg(color.getColor()).a(BOLA).reset().toString();
    }


}

File: Urnas-y-Bolas/src/Bola.kt
import org.fusesource.jansi.Ansi
import org.fusesource.jansi.AnsiConsole

class Bola(private val color: Color) {
    companion object {
        private const val BOLA = '\u25CF'  // Representación textual
        private var contador = 0           // Nº de bolas que existen

        // Método para generar una bola con un color aleatorio
        fun generar(): Bola {
            val colores = Color.values()
            val n = (Math.random() * colores.size).toInt()
            return Bola(colores[n])
        }

        // Método para obtener el total de bolas creadas
        fun totalBolas(): Int {
            return contador
        }
    }

    val numero: Int
    private val color: Color

    init {
        numero = ++contador
        this.color = color
    }

    // Métodos de acceso
    fun getColor(): Color {
        return color
    }


    fun getNumero(): Int {
        return numero
    }

    // Representación en texto de la bola con color
    override fun toString(): String {
        return Ansi.ansi().fg(color.getColor()).a(BOLA).reset().toString()
    }
}
